File: rust/autonomi/src/archive.rs
//! Archive module - File collections with metadata for the Autonomi network
//!
//! ## Current Implementation
//! - ✅ Metadata: File metadata (size, creation time, modification time)
//! - ✅ PublicArchive: Collection of public files with metadata
//! - ✅ PrivateArchive: Collection of private files with encryption
//! - ✅ ArchiveAddress: Address for public archives on the network
//! - ✅ PrivateArchiveDataMap: Datamap for private archives
//! - ✅ Client methods: archive_cost, archive_get, archive_put, archive_put_public, archive_get_public

use autonomi::files::archive_private::PrivateArchiveDataMap as AutonomiPrivateArchiveDataMap;
use autonomi::files::archive_public::ArchiveAddress as AutonomiArchiveAddress;
use autonomi::files::{
    Metadata as AutonomiMetadata, PrivateArchive as AutonomiPrivateArchive,
    PublicArchive as AutonomiPublicArchive,
};
use std::sync::Arc;

use crate::data::{DataAddress, DataMapChunk};

/// Error type for archive operations
#[derive(Debug, uniffi::Error, thiserror::Error)]
pub enum ArchiveError {
    #[error("Invalid archive: {reason}")]
    InvalidArchive { reason: String },
    #[error("Parsing failed: {reason}")]
    ParsingFailed { reason: String },
    #[error("File not found: {path}")]
    FileNotFound { path: String },
}

/// Metadata for files in an archive, containing creation time, modification time, and size.
#[derive(uniffi::Object, Clone, Debug)]
pub struct Metadata {
    pub(crate) inner: AutonomiMetadata,
}

#[uniffi::export]
impl Metadata {
    /// Create new metadata with the given file size.
    /// Creation and modification times are set to the current time.
    #[uniffi::constructor]
    pub fn new(size: u64) -> Arc<Self> {
        Arc::new(Self {
            inner: AutonomiMetadata::new_with_size(size),
        })
    }

    /// Create metadata with specific timestamps
    #[uniffi::constructor]
    pub fn with_timestamps(size: u64, created: u64, modified: u64) -> Arc<Self> {
        Arc::new(Self {
            inner: AutonomiMetadata {
                size,
                created,
                modified,
                extra: None,
            },
        })
    }

    /// Get the file size in bytes
    pub fn size(&self) -> u64 {
        self.inner.size
    }

    /// Get the creation time as Unix timestamp in seconds
    pub fn created(&self) -> u64 {
        self.inner.created
    }

    /// Get the modification time as Unix timestamp in seconds
    pub fn modified(&self) -> u64 {
        self.inner.modified
    }
}

/// Address of a public archive on the network
#[derive(uniffi::Object, Clone, Copy, Debug)]
pub struct ArchiveAddress {
    pub(crate) inner: AutonomiArchiveAddress,
}

#[uniffi::export]
impl ArchiveAddress {
    /// Create an ArchiveAddress from a hex string
    #[uniffi::constructor]
    pub fn from_hex(hex: String) -> Result<Arc<Self>, ArchiveError> {
        let inner =
            AutonomiArchiveAddress::from_hex(&hex).map_err(|e| ArchiveError::ParsingFailed {
                reason: format!("Failed to parse hex: {}", e),
            })?;
        Ok(Arc::new(Self { inner }))
    }

    /// Returns the hex string representation of this archive address
    pub fn to_hex(&self) -> String {
        self.inner.to_hex()
    }
}

/// Datamap for a private archive, used to retrieve the archive from the network
#[derive(uniffi::Object, Clone, Debug)]
pub struct PrivateArchiveDataMap {
    pub(crate) inner: AutonomiPrivateArchiveDataMap,
}

#[uniffi::export]
impl PrivateArchiveDataMap {
    /// Create a PrivateArchiveDataMap from a hex string
    #[uniffi::constructor]
    pub fn from_hex(hex: String) -> Result<Arc<Self>, ArchiveError> {
        let inner = AutonomiPrivateArchiveDataMap::from_hex(&hex).map_err(|e| {
            ArchiveError::ParsingFailed {
                reason: format!("Failed to parse hex: {}", e),
            }
        })?;
        Ok(Arc::new(Self { inner }))
    }

    /// Returns the hex string representation of this private archive datamap
    pub fn to_hex(&self) -> String {
        self.inner.to_hex()
    }
}

/// A file entry in a public archive
#[derive(uniffi::Record)]
pub struct PublicArchiveFileEntry {
    /// The path of the file in the archive
    pub path: String,
    /// The data address where the file content is stored
    pub address: Arc<DataAddress>,
    /// Metadata about the file
    pub metadata: Arc<Metadata>,
}

/// A file entry in a private archive
#[derive(uniffi::Record)]
pub struct PrivateArchiveFileEntry {
    /// The path of the file in the archive
    pub path: String,
    /// The data map chunk to retrieve the file content
    pub data_map: Arc<DataMapChunk>,
    /// Metadata about the file
    pub metadata: Arc<Metadata>,
}

/// A public archive containing files that can be accessed by anyone on the network.
#[derive(uniffi::Object, Clone, Debug)]
pub struct PublicArchive {
    pub(crate) inner: AutonomiPublicArchive,
}

#[uniffi::export]
impl PublicArchive {
    /// Create a new empty public archive
    #[uniffi::constructor]
    pub fn new() -> Arc<Self> {
        Arc::new(Self {
            inner: AutonomiPublicArchive::new(),
        })
    }

    /// Add a file to the archive
    pub fn add_file(
        &self,
        path: String,
        address: Arc<DataAddress>,
        metadata: Arc<Metadata>,
    ) -> Arc<Self> {
        let mut archive = self.inner.clone();
        archive.add_file(
            std::path::PathBuf::from(path),
            address.inner,
            metadata.inner.clone(),
        );
        Arc::new(Self { inner: archive })
    }

    /// Rename a file in the archive
    pub fn rename_file(
        &self,
        old_path: String,
        new_path: String,
    ) -> Result<Arc<Self>, ArchiveError> {
        let mut archive = self.inner.clone();
        archive
            .rename_file(
                &std::path::PathBuf::from(&old_path),
                &std::path::PathBuf::from(&new_path),
            )
            .map_err(|e| ArchiveError::InvalidArchive {
                reason: format!("Failed to rename file: {}", e),
            })?;
        Ok(Arc::new(Self { inner: archive }))
    }

    /// List all files in the archive
    pub fn files(&self) -> Vec<PublicArchiveFileEntry> {
        self.inner
            .map()
            .iter()
            .map(|(path, (addr, meta))| PublicArchiveFileEntry {
                path: path.to_string_lossy().to_string(),
                address: Arc::new(DataAddress { inner: *addr }),
                metadata: Arc::new(Metadata {
                    inner: meta.clone(),
                }),
            })
            .collect()
    }

    /// Get the number of files in the archive
    pub fn file_count(&self) -> u64 {
        self.inner.map().len() as u64
    }

    /// Get all data addresses in the archive
    pub fn addresses(&self) -> Vec<String> {
        self.inner
            .addresses()
            .into_iter()
            .map(|a| a.to_hex())
            .collect()
    }
}

/// A private archive containing files with encrypted access.
#[derive(uniffi::Object, Clone, Debug)]
pub struct PrivateArchive {
    pub(crate) inner: AutonomiPrivateArchive,
}

#[uniffi::export]
impl PrivateArchive {
    /// Create a new empty private archive
    #[uniffi::constructor]
    pub fn new() -> Arc<Self> {
        Arc::new(Self {
            inner: AutonomiPrivateArchive::new(),
        })
    }

    /// Add a file to the archive
    pub fn add_file(
        &self,
        path: String,
        data_map: Arc<DataMapChunk>,
        metadata: Arc<Metadata>,
    ) -> Arc<Self> {
        let mut archive = self.inner.clone();
        archive.add_file(
            std::path::PathBuf::from(path),
            data_map.inner.clone(),
            metadata.inner.clone(),
        );
        Arc::new(Self { inner: archive })
    }

    /// Rename a file in the archive
    pub fn rename_file(
        &self,
        old_path: String,
        new_path: String,
    ) -> Result<Arc<Self>, ArchiveError> {
        let mut archive = self.inner.clone();
        archive
            .rename_file(
                &std::path::PathBuf::from(&old_path),
                &std::path::PathBuf::from(&new_path),
            )
            .map_err(|e| ArchiveError::InvalidArchive {
                reason: format!("Failed to rename file: {}", e),
            })?;
        Ok(Arc::new(Self { inner: archive }))
    }

    /// List all files in the archive
    pub fn files(&self) -> Vec<PrivateArchiveFileEntry> {
        self.inner
            .map()
            .iter()
            .map(|(path, (data_map, meta))| PrivateArchiveFileEntry {
                path: path.to_string_lossy().to_string(),
                data_map: Arc::new(DataMapChunk {
                    inner: data_map.clone(),
                }),
                metadata: Arc::new(Metadata {
                    inner: meta.clone(),
                }),
            })
            .collect()
    }

    /// Get the number of files in the archive
    pub fn file_count(&self) -> u64 {
        self.inner.map().len() as u64
    }

    /// Get all data maps in the archive
    pub fn data_maps(&self) -> Vec<Arc<DataMapChunk>> {
        self.inner
            .data_maps()
            .into_iter()
            .map(|dm| Arc::new(DataMapChunk { inner: dm }))
            .collect()
    }
}

File: csharp/AntFfi/Archive.cs
using AntFfi.Native;

namespace AntFfi;

/// <summary>
/// Exception thrown when an archive operation fails.
/// </summary>
public class ArchiveException : AntFfiException
{
    public ArchiveException(string message) : base(message) { }
    public ArchiveException(string message, RustCallStatus status) : base(message, status) { }
}

/// <summary>
/// File metadata containing size and timestamps.
/// </summary>
public sealed class Metadata : NativeHandle
{
    internal Metadata(IntPtr handle) : base(handle) { }

    /// <summary>
    /// Creates new metadata with the given file size.
    /// Creation and modification times are set to the current time.
    /// </summary>
    /// <param name="size">File size in bytes.</param>
    public static Metadata Create(ulong size)
    {
        var status = RustCallStatus.Create();
        var handle = NativeMethods.MetadataNew(size, ref status);
        UniFFIHelpers.CheckStatus(ref status, "Metadata.Create");
        return new Metadata(handle);
    }

    /// <summary>
    /// Creates metadata with specific timestamps.
    /// </summary>
    /// <param name="size">File size in bytes.</param>
    /// <param name="created">Creation time as Unix timestamp in seconds.</param>
    /// <param name="modified">Modification time as Unix timestamp in seconds.</param>
    public static Metadata WithTimestamps(ulong size, ulong created, ulong modified)
    {
        var status = RustCallStatus.Create();
        var handle = NativeMethods.MetadataWithTimestamps(size, created, modified, ref status);
        UniFFIHelpers.CheckStatus(ref status, "Metadata.WithTimestamps");
        return new Metadata(handle);
    }

    /// <summary>
    /// Gets the file size in bytes.
    /// </summary>
    public ulong Size
    {
        get
        {
            ThrowIfDisposed();
            var status = RustCallStatus.Create();
            return NativeMethods.MetadataSize(CloneHandle(), ref status);
        }
    }

    /// <summary>
    /// Gets the creation time as Unix timestamp in seconds.
    /// </summary>
    public ulong Created
    {
        get
        {
            ThrowIfDisposed();
            var status = RustCallStatus.Create();
            return NativeMethods.MetadataCreated(CloneHandle(), ref status);
        }
    }

    /// <summary>
    /// Gets the modification time as Unix timestamp in seconds.
    /// </summary>
    public ulong Modified
    {
        get
        {
            ThrowIfDisposed();
            var status = RustCallStatus.Create();
            return NativeMethods.MetadataModified(CloneHandle(), ref status);
        }
    }

    /// <inheritdoc/>
    protected override void FreeHandle()
    {
        var status = RustCallStatus.Create();
        NativeMethods.FreeMetadata(Handle, ref status);
    }

    /// <inheritdoc/>
    protected internal override IntPtr CloneHandle()
    {
        var status = RustCallStatus.Create();
        return NativeMethods.CloneMetadata(Handle, ref status);
    }
}

/// <summary>
/// Address of a public archive on the network.
/// </summary>
public sealed class ArchiveAddress : NativeHandle
{
    internal ArchiveAddress(IntPtr handle) : base(handle) { }

    /// <summary>
    /// Creates an ArchiveAddress from a hex string.
    /// </summary>
    public static ArchiveAddress FromHex(string hex)
    {
        var hexBuffer = UniFFIHelpers.StringToRustBuffer(hex);
        var status = RustCallStatus.Create();
        var handle = NativeMethods.ArchiveAddressFromHex(hexBuffer, ref status);
        if (status.IsError)
            throw new ArchiveException("Failed to parse archive address from hex", status);
        return new ArchiveAddress(handle);
    }

    /// <summary>
    /// Returns the hex representation of this archive address.
    /// </summary>
    public string ToHex()
    {
        ThrowIfDisposed();
        var status = RustCallStatus.Create();
        NativeMethods.ArchiveAddressToHex(out var buffer, CloneHandle(), ref status);
        UniFFIHelpers.CheckStatus(ref status, "ArchiveAddress.ToHex");
        return UniFFIHelpers.StringFromRustBuffer(buffer);
    }

    /// <inheritdoc/>
    protected override void FreeHandle()
    {
        var status = RustCallStatus.Create();
        NativeMethods.FreeArchiveAddress(Handle, ref status);
    }

    /// <inheritdoc/>
    protected internal override IntPtr CloneHandle()
    {
        var status = RustCallStatus.Create();
        return NativeMethods.CloneArchiveAddress(Handle, ref status);
    }
}

/// <summary>
/// Datamap for a private archive, used to retrieve the archive from the network.
/// </summary>
public sealed class PrivateArchiveDataMap : NativeHandle
{
    internal PrivateArchiveDataMap(IntPtr handle) : base(handle) { }

    /// <summary>
    /// Creates a PrivateArchiveDataMap from a hex string.
    /// </summary>
    public static PrivateArchiveDataMap FromHex(string hex)
    {
        var hexBuffer = UniFFIHelpers.StringToRustBuffer(hex);
        var status = RustCallStatus.Create();
        var handle = NativeMethods.PrivateArchiveDataMapFromHex(hexBuffer, ref status);
        if (status.IsError)
            throw new ArchiveException("Failed to parse private archive datamap from hex", status);
        return new PrivateArchiveDataMap(handle);
    }

    /// <summary>
    /// Returns the hex representation of this private archive datamap.
    /// </summary>
    public string ToHex()
    {
        ThrowIfDisposed();
        var status = RustCallStatus.Create();
        NativeMethods.PrivateArchiveDataMapToHex(out var buffer, CloneHandle(), ref status);
        UniFFIHelpers.CheckStatus(ref status, "PrivateArchiveDataMap.ToHex");
        return UniFFIHelpers.StringFromRustBuffer(buffer);
    }

    /// <inheritdoc/>
    protected override void FreeHandle()
    {
        var status = RustCallStatus.Create();
        NativeMethods.FreePrivateArchiveDataMap(Handle, ref status);
    }

    /// <inheritdoc/>
    protected internal override IntPtr CloneHandle()
    {
        var status = RustCallStatus.Create();
        return NativeMethods.ClonePrivateArchiveDataMap(Handle, ref status);
    }
}

/// <summary>
/// A public archive containing files that can be accessed by anyone on the network.
/// </summary>
/// <remarks>
/// PublicArchive is immutable - operations like AddFile return a new archive.
/// </remarks>
public sealed class PublicArchive : NativeHandle
{
    internal PublicArchive(IntPtr handle) : base(handle) { }

    /// <summary>
    /// Creates a new empty public archive.
    /// </summary>
    public static PublicArchive Create()
    {
        var status = RustCallStatus.Create();
        var handle = NativeMethods.PublicArchiveNew(ref status);
        UniFFIHelpers.CheckStatus(ref status, "PublicArchive.Create");
        return new PublicArchive(handle);
    }

    /// <summary>
    /// Adds a file to the archive and returns a new archive.
    /// </summary>
    /// <param name="path">The path of the file in the archive.</param>
    /// <param name="address">The data address where the file content is stored.</param>
    /// <param name="metadata">Metadata about the file.</param>
    /// <returns>A new archive with the file added.</returns>
    public PublicArchive AddFile(string path, DataAddress address, Metadata metadata)
    {
        ThrowIfDisposed();
        ArgumentNullException.ThrowIfNull(address);
        ArgumentNullException.ThrowIfNull(metadata);
        address.ThrowIfDisposed();
        metadata.ThrowIfDisposed();

        var pathBuffer = UniFFIHelpers.StringToRustBuffer(path);
        var status = RustCallStatus.Create();
        var handle = NativeMethods.PublicArchiveAddFile(CloneHandle(), pathBuffer, address.CloneHandle(), metadata.CloneHandle(), ref status);
        UniFFIHelpers.CheckStatus(ref status, "PublicArchive.AddFile");
        return new PublicArchive(handle);
    }

    /// <summary>
    /// Renames a file in the archive and returns a new archive.
    /// </summary>
    public PublicArchive RenameFile(string oldPath, string newPath)
    {
        ThrowIfDisposed();

        var oldPathBuffer = UniFFIHelpers.StringToRustBuffer(oldPath);
        var newPathBuffer = UniFFIHelpers.StringToRustBuffer(newPath);
        var status = RustCallStatus.Create();
        var handle = NativeMethods.PublicArchiveRenameFile(CloneHandle(), oldPathBuffer, newPathBuffer, ref status);
        if (status.IsError)
            throw new ArchiveException("Failed to rename file in archive", status);
        return new PublicArchive(handle);
    }

    /// <summary>
    /// Gets the number of files in the archive.
    /// </summary>
    public ulong FileCount
    {
        get
        {
            ThrowIfDisposed();
            var status = RustCallStatus.Create();
            return NativeMethods.PublicArchiveFileCount(CloneHandle(), ref status);
        }
    }

    /// <inheritdoc/>
    protected override void FreeHandle()
    {
        var status = RustCallStatus.Create();
        NativeMethods.FreePublicArchive(Handle, ref status);
    }

    /// <inheritdoc/>
    protected internal override IntPtr CloneHandle()
    {
        var status = RustCallStatus.Create();
        return NativeMethods.ClonePublicArchive(Handle, ref status);
    }
}

/// <summary>
/// A private archive containing files with encrypted access.
/// </summary>
/// <remarks>
/// PrivateArchive is immutable - operations like AddFile return a new archive.
/// </remarks>
public sealed class PrivateArchive : NativeHandle
{
    internal PrivateArchive(IntPtr handle) : base(handle) { }

    /// <summary>
    /// Creates a new empty private archive.
    /// </summary>
    public static PrivateArchive Create()
    {
        var status = RustCallStatus.Create();
        var handle = NativeMethods.PrivateArchiveNew(ref status);
        UniFFIHelpers.CheckStatus(ref status, "PrivateArchive.Create");
        return new PrivateArchive(handle);
    }

    /// <summary>
    /// Adds a file to the archive and returns a new archive.
    /// </summary>
    /// <param name="path">The path of the file in the archive.</param>
    /// <param name="dataMap">The data map chunk to retrieve the file content.</param>
    /// <param name="metadata">Metadata about the file.</param>
    /// <returns>A new archive with the file added.</returns>
    public PrivateArchive AddFile(string path, DataMapChunk dataMap, Metadata metadata)
    {
        ThrowIfDisposed();
        ArgumentNullException.ThrowIfNull(dataMap);
        ArgumentNullException.ThrowIfNull(metadata);
        dataMap.ThrowIfDisposed();
        metadata.ThrowIfDisposed();

        var pathBuffer = UniFFIHelpers.StringToRustBuffer(path);
        var status = RustCallStatus.Create();
        var handle = NativeMethods.PrivateArchiveAddFile(CloneHandle(), pathBuffer, dataMap.CloneHandle(), metadata.CloneHandle(), ref status);
        UniFFIHelpers.CheckStatus(ref status, "PrivateArchive.AddFile");
        return new PrivateArchive(handle);
    }

    /// <summary>
    /// Renames a file in the archive and returns a new archive.
    /// </summary>
    public PrivateArchive RenameFile(string oldPath, string newPath)
    {
        ThrowIfDisposed();

        var oldPathBuffer = UniFFIHelpers.StringToRustBuffer(oldPath);
        var newPathBuffer = UniFFIHelpers.StringToRustBuffer(newPath);
        var status = RustCallStatus.Create();
        var handle = NativeMethods.PrivateArchiveRenameFile(CloneHandle(), oldPathBuffer, newPathBuffer, ref status);
        if (status.IsError)
            throw new ArchiveException("Failed to rename file in archive", status);
        return new PrivateArchive(handle);
    }

    /// <summary>
    /// Gets the number of files in the archive.
    /// </summary>
    public ulong FileCount
    {
        get
        {
            ThrowIfDisposed();
            var status = RustCallStatus.Create();
            return NativeMethods.PrivateArchiveFileCount(CloneHandle(), ref status);
        }
    }

    /// <inheritdoc/>
    protected override void FreeHandle()
    {
        var status = RustCallStatus.Create();
        NativeMethods.FreePrivateArchive(Handle, ref status);
    }

    /// <inheritdoc/>
    protected internal override IntPtr CloneHandle()
    {
        var status = RustCallStatus.Create();
        return NativeMethods.ClonePrivateArchive(Handle, ref status);
    }
}
